### ISSUE TITLE
Edge Cases Generation Order

### ISSUE THREAD

**Issue by luvarqpp** [jlink/jqwik#101]
## Testing Problem

Order of edge cases does matter. When there are too many edge cases (for example due complex arbitrary like `Combinators.combine(... many arbitraries ...)` ), there are two things:

* when there are more edge cases than all Property tries (by default 1000), not all edge cases will be tried and no other case will be tried. It will prevent any random test case (when `defaultEdgeCases=FIRST` is set)
* order of edge cases does mater and does not effectively use resources used during test. Edge cases should be used in random order (at least while `defaultEdgeCases=MIXIN` is set).

## Suggested Solution

There were two types of problems. I suggest mitigation like this:

* print warning to log/console when there is not enough tries to cover all edge cases. Make able to suppress this warning (or perhaps print it only when more than 10 test runs are needed to have 99% chance of covering all edge cases?)
* use edge cases in random order. No mater if `defaultEdgeCases=FIRST`or `defaultEdgeCases=MIXIN` is in use

## Discussion

I have no extensive experience with edge cases in wild. I have one more note for them through (third thing in this single issue). When `defaultEdgeCases=MIXIN` is in use, how often are edge cases mixed in? I would suggest to insert than in such way, that after certain test runs (10 test runs?), all test cases will be used (with some probability; or could we exploit `.jqwik-database` file to try them in round-robin manner?)

**Comment by jlink** [comment]
I'll probably start with a small step: Reporting the number of tried and overall edge-cases:

```
                              |-------------------jqwik-------------------
tries = 100                   | # of calls to property
...
edge-cases = MIXIN (87/220)   | edge cases are mixed in (#tried/#all)
```

Mixing them in randomly is further down the road.

As for the mix-in probability, the ratio is calculated like this:

	private static int calculateBaseToEdgeCaseRatio(int countEdgeCases, int genSize) {
		return min(
			max(genSize / countEdgeCases, 3) - 1,
			10
		);
	}

Thus, the ratio is between 1:3 and 1:10 until all edge cases have been used up.

**Comment by jlink** [comment]
I'm currently not sure how difficult random order of edge cases would be. The problem is that edge cases are generated on the fly because their number can potentially be very large. Shuffling them would require to generate all of them first which might take a long time and slow-down performance.

**Comment by luvarqpp** [comment]
> I'm currently not sure how difficult random order of edge cases would be. The problem is that edge cases are generated on the fly because their number can potentially be very large. Shuffling them would require to generate all of them first which might take a long time and slow-down performance.

Is it possible, to generate "next" edge case in random manner easily? In single attribute (arbitrary) I think there will be no problem. Through in combined ones and multiple parameters for test methods can be non-trivial.

Can you pinpoint some code, where edge cases are actually generated?

**Comment by jlink** [comment]
> > I'm currently not sure how difficult random order of edge cases would be. The problem is that edge cases are generated on the fly because their number can potentially be very large. Shuffling them would require to generate all of them first which might take a long time and slow-down performance.
> 
> Is it possible, to generate "next" edge case in random manner easily? In single attribute (arbitrary) I think there will be no problem. Through in combined ones and multiple parameters for test methods can be non-trivial.
> 
> Can you pinpoint some code, where edge cases are actually generated?

Yes, single parameter property would be easy - but mostly worthless.

The magic code is in https://github.com/jlink/jqwik/blob/5217b8e8d6fc7379c08ab077354c1989da9a16ea/engine/src/main/java/net/jqwik/engine/properties/EdgeCasesGenerator.java#L27
It combines several iterators to a single one. But shuffling the iterators internally does not really help because the order of values withing the iterators will always be the same, otherwise one wouldn't know that all cases have been generated.

**Comment by luvarqpp** [comment]
> ```
>                               |-------------------jqwik-------------------
> tries = 100                   | # of calls to property
> ...
> edge-cases = MIXIN (87/220)   | edge cases are mixed in (#tried/#all)
> ```

I would consider separating edge-cases usage and their statistics. i.e.:

```
|-------------------jqwik-------------------
tries = 100                   | # of calls to property
...
edge-cases = MIXIN (87/220)   | edge cases are mixed in (#tried/#all)
edge-cases-total = 220        | number of all possible edge cases for this test (combined fro all arbitraries)
edge-cases-used = 87          | edge cases that has been passed to some try actually
```

Just to keep output format in `key = value` format. On the other side, there will be more statistics lines than I can count on my hand...

Is there any mechanism, that will affect statistics? i.e. when I change edge-cases generation from default, perhaps I would like see all extra statistics for this matter. When there is default (and no override in `@Property`, hide given statistics line).

**Comment by jlink** [comment]
How about:

```
|-------------------jqwik-------------------
tries = 100                      | # of calls to property
...
edge-cases#mode = MIXIN (87/220) | edge cases are mixed in (#tried/#all)
edge-cases#total = 220           | # of all combined possible edge cases
edge-cases#tried = 87            | # of edge cases that have been tried in current run

```

**Comment by jlink** [comment]
> 
> Is there any mechanism, that will affect statistics? i.e. when I change edge-cases generation from default, perhaps I would like see all extra statistics for this matter. When there is default (and no override in `@Property`, hide given statistics line).

What statistics do you mean. Edge case statistics or general statistics invoked through `Statistics.collect(..)`?

**Comment by luvarqpp** [comment]
> How about:
> 
> ```
> |-------------------jqwik-------------------
> tries = 100                      | # of calls to property
> ...
> edge-cases#mode = MIXIN (87/220) | edge cases are mixed in (#tried/#all)
> edge-cases#total = 220           | # of all combined possible edge cases
> edge-cases#tried = 87            | # of edge cases that have been tried in current run
> ```

Seems fine to me.


> > Is there any mechanism, that will affect statistics? i.e. when I change edge-cases generation from default, perhaps I would like see all extra statistics for this matter. When there is default (and no override in `@Property`, hide given statistics line).
> 
> What statistics do you mean. Edge case statistics or general statistics invoked through `Statistics.collect(..)`?

I mean edge case statistics and generally things that are included in "table" printed after finishing property test.

**Comment by jlink** [comment]
> > > Is there any mechanism, that will affect statistics? i.e. when I change edge-cases generation from default, perhaps I would like see all extra statistics for this matter. When there is default (and no override in `@Property`, hide given statistics line).
> > 
> > 
> > What statistics do you mean. Edge case statistics or general statistics invoked through `Statistics.collect(..)`?
> 
> I mean edge case statistics and generally things that are included in "table" printed after finishing property test.

I could show/hide those statistics depending on explicit choice of edge-cases mode. I wonder if it would confuse more than help if those values sometimes show up and sometimes don't. The one obvious thing is to leave the numbers out when `EdgeCasesMode.NONE` is chosen.

**Comment by luvarqpp** [comment]
Confusion can happen... What about to keep only some "normal" statistics printed out by default and introducing mechanism like this?:

```
@Property(tries = 1300)
@Report({Reporting.GENERATED, Reporting.EDGE_CASES_DETAILS, Reporting.RUNTIME_DETAILS})
public void testUserValidity(@ForAll User user() {};
```

**Comment by jlink** [comment]
What you can do is switching reporting off for successful properties in `jqwik.properties`.

What's your pain with having all things reported that can be reported? Are reports too long, too confusing?

I'm trying to figure out if time spent for making all this stuff configurable is well spent.

**Comment by luvarqpp** [comment]
Technically reporting always all things (or nothing, when turned reporting off) is imho OK. I was perhaps thinking in a "premature optimization" style to include interesting report parts only in certain properties (tests), where they are used.

To keep things simple, I am OK with "report everything or nothing" style.

**Comment by jlink** [comment]
Edge case reporting implemented in https://github.com/jlink/jqwik/commit/071868043a341e4516fbe6154ae6b5c59bcd19cd

Available in 1.3.0-SNAPSHOT

**Comment by jlink** [comment]
After revisiting and reimplementing how edge cases are currently being generated this issue is too far from implementation to make sense keeping it open. Closing it.
